File: backend/Common/Permissions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using System.Threading.Tasks;
using backend.Models;

namespace backend.Common
{
    public class Permissions
    {
        public const string All = "*";
        public const string UsuariosManage = "usuarios:manage";
        public const string EmpleadosRead = "empleados:read";
        public const string EmpleadosCreate = "empleados:create";
        public const string EmpleadosEdit = "empleados:edit";
        public const string EmpleadosDelete = "empleados:delete";
        public const string NominaRead = "nomina:read";
        public const string NominaProcess = "nomina:process";
        public const string VolantesSend = "volantes:send";
        public const string EventosManage = "eventos:manage";

        public static readonly Dictionary<string, string[]> DefaultRolePermissions = new(StringComparer.OrdinalIgnoreCase)
        {
          ["Admin"] = new[] { All },
          ["RRHH"] = new[]
        {
            EmpleadosRead,
            EmpleadosCreate,
            EmpleadosEdit,
            EmpleadosDelete,
            NominaRead,
            NominaProcess,
            VolantesSend,
            EventosManage
        },

        ["Contador"] = new[]
        {
          EmpleadosRead,
          NominaRead,
          NominaProcess,
          VolantesSend,
          EventosManage

        },

        ["Auditor"] = new[]
        {
          EmpleadosRead,
          NominaRead,
        },

        ["Operador"] = new[]
        {
          EmpleadosRead,
          EmpleadosCreate,
          EmpleadosEdit,
          NominaRead
        }
        };

        public static string[] ObtenerPermisosPorRol(string? rol)
        {
            if (string.IsNullOrWhiteSpace(rol)) return new[] {EmpleadosRead};

            return DefaultRolePermissions.TryGetValue(rol.Trim(), out var permisos)
            ? permisos
            : new[] {EmpleadosRead}; 
        }

        public static bool TienePermiso(Usuario? usuario, string permisoRequerido)
        {
            if (usuario == null || !usuario.Activo) return false;

            if (string.Equals(usuario.Rol, "Admin", StringComparison.OrdinalIgnoreCase)) return true;

            var permisosEfectivos = new HashSet<string>(ObtenerPermisosPorRol(usuario.Rol), StringComparer.OrdinalIgnoreCase);
            
            if (!string.IsNullOrWhiteSpace(usuario.PermisosJson) && usuario.PermisosJson != "[]")
            {
                try
                {
                    var customPerms = JsonSerializer.Deserialize<List<string>>(usuario.PermisosJson);

                    if (customPerms != null)
                    {
                        foreach (var perm in customPerms)
                        {
                            permisosEfectivos.Add(perm);
                        }
                    }
                }

                catch (JsonException)
                {
                    // Manejar error de deserialización si es necesario
                }
            }

            return permisosEfectivos.Contains(All) || permisosEfectivos.Contains(permisoRequerido);
        }
    }
}

File: sistema-nomina/src/constants/permissions.ts
import { Usuario } from '../types/usuario';

export type Permission =
  | '*'
  | 'usuarios:manage'
  | 'empleados:read'
  | 'empleados:create'
  | 'empleados:edit'
  | 'empleados:delete'
  | 'nomina:read'
  | 'nomina:process'
  | 'volantes:send'
  | 'eventos:manage';

export const DEFAULT_ROLE_PERMISSIONS: Record<string, Permission[]> = {
  Admin: ['*'],
  RRHH: [
    'empleados:read',
    'empleados:create',
    'empleados:edit',
    'empleados:delete',
    'nomina:read',
    'nomina:process',
    'volantes:send',
    'eventos:manage',
  ],
  Contador: [
    'empleados:read',
    'nomina:read',
    'nomina:process',
    'volantes:send',
    'eventos:manage',
  ],
  Auditor: [
    'empleados:read',
    'nomina:read',
  ],
    Operador: [
    'empleados:read',
    'empleados:create',
    'empleados:edit',
    'nomina:read',
  ],
};

/**
 * Obtiene los permisos efectivos para un rol determinado.
 */
export const getPermissionsForRole = (rol: string): Permission[] => {
  const normalizedRole = Object.keys(DEFAULT_ROLE_PERMISSIONS).find(
    (key) => key.toLowerCase() === rol.trim().toLowerCase()
  );
  return normalizedRole ? DEFAULT_ROLE_PERMISSIONS[normalizedRole] : ['empleados:read'];
};

/**
 * Parsea de forma segura el JSON de permisos de un usuario o devuelve los permisos predeterminados de su rol.
 */
export const getUserPermissions = (user: Usuario | null | undefined): string[] => {
  if (!user) return [];

  const defaultRolePerms = getPermissionsForRole(user.rol);
  let customPerms: string[] = [];

  if (user.permisosJson && user.permisosJson.trim() !== '' && user.permisosJson !== '[]') {
    try {
      const parsed = JSON.parse(user.permisosJson);
      if (Array.isArray(parsed)) {
        customPerms = parsed;
      }
    } catch {
      // Ignorar error de parseo
    }
  }

  // Unir los permisos del rol con los permisos del JSON (sin duplicados)
  return Array.from(new Set([...defaultRolePerms, ...customPerms]));
};

/**
 * Comprueba si un usuario tiene un permiso específico o acceso total (*).
 */
export const hasPermission = (user: Usuario | null | undefined, permission: Permission): boolean => {
  if (!user || !user.activo) return false;
  if (user.rol.toUpperCase() === 'ADMIN') return true;

  const userPerms = getUserPermissions(user);
  return userPerms.includes('*') || userPerms.includes(permission);
};

/**
 * Comprueba si un usuario tiene al menos uno de los permisos indicados.
 */
export const hasAnyPermission = (user: Usuario | null | undefined, permissions: Permission[]): boolean => {
  if (!user || !user.activo) return false;
  if (user.rol.toUpperCase() === 'ADMIN') return true;

  const userPerms = getUserPermissions(user);
  if (userPerms.includes('*')) return true;

  return permissions.some((p) => userPerms.includes(p));
};
